### PR TITLE
refactor(sdk): Refactor Report API and include limited json validation

### PR DESCRIPTION
**Pull request by andrewtruong**
Description
-----------
This PR makes mostly internal improvements to Report API to make it easier to manage.

Notable changes:
1. Decouples the interface (`interface.py`) from the backend (`internal.py`), where previously the json was modified directly(!)
2. Adds better and faster validation via pydantic
3. Adds support for collapsible headings

This does not introduce any breaking changes to working v1 Reports.
1. A new v2 namespace is introduced, and a deprecation warning is displayed to v1 users
2. The v1 reports will be phased out in future and replaced with v2.

Testing
-------
- unit tests
- e2e test in core

### CODE REVIEW

**Review by ssisk — COMMENTED** [review]
I haven't fully looked at this, but after discussing with andrew I think this is a good plan, thanks!

File: wandb/apis/reports2/__init__.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+def show_welcome_message():
+    if os.getenv("WANDB_DISABLE_REPORT_API_MESSAGE"):
+        return
+
+    termlog(
+        cleandoc(
+            """
+            Thanks for trying out Report API v2!
+            See a tutorial and the changes here: https://colab.research.google.com/drive/1CzyJx1nuOS4pdkXa2XPaRQyZdmFmLmXV
```
Comment (vanpelt) [inline wandb/apis/reports2/__init__.py:31]: Would be smart to replace this with a short link at wandb.me.

**Review by vanpelt — APPROVED** [review]
Looks good, I'm not a big fan of `reports2`.  `reportsv2` feels a little better.  Another option would be to put it under reports `reports.v2` then eventually move the current reports to `reports.v1` and make `reports` the new stuff.  Don't feel strong enough to block so do as you please.

File: requirements_test.txt [hunk 2]
```diff
@@ -15,3 +15,5 @@ pandas
 responses
 
 coverage[toml]
+pydantic>=2.0.0
```
Comment (kptkin) [inline requirements_test.txt:18]: why here if you already put it in dev?

File: requirements_dev.txt [hunk 3]
```diff
@@ -43,7 +43,8 @@ responses
 respx
 prometheus_client
 google-cloud-aiplatform
-
+pydantic>=2.0.0
```
Comment (kptkin) [inline requirements_dev.txt:46]: why do you need it here if you put it in requirements?

File: requirements.txt [hunk 4]
```diff
@@ -14,3 +14,4 @@ setuptools
 appdirs>=1.4.3
 dataclasses; python_version < '3.7'
 typing_extensions; python_version < '3.10'
+pydantic>=2.0.0
```
Comment (kptkin) [inline requirements.txt:17]: also maybe make it an extra for reports?

File: setup.py [hunk 5]
```diff
@@ -1,3 +1,3 @@
 from setuptools import setup
 
-setup()
+setup(extras_require={"reports": ["pydantic>=2.0.0"]})
```
Comment (andrewtruong) [inline setup.py:3]: Is this the right place for extras @kptkin ?

Comment (vanpelt) [reply setup.py:3]: `pyproject.toml` => [project.optional-dependencies]

File: requirements_test.txt [hunk 6]
```diff
@@ -15,3 +15,6 @@ pandas
 responses
 
 coverage[toml]
+polyfactory; python_version >= '3.8'
+
+.[reports]
```
Comment (dmitryduev) [inline requirements_test.txt:20]: I think you don't need this, since you already put these into _dev

File: setup.py (hunk 5 above)
Comment (kptkin) [reply setup.py]: like @vanpelt said, you can see other extras for examples.

File: requirements_test.txt (hunk 6 above)
Comment (dmitryduev) [inline requirements_test.txt:20]:
```suggestion

```

File: .circleci/config.yml [hunk 7]
```diff
@@ -1763,6 +1763,19 @@ workflows:
...
           xdist: 1
 
+      # report api tests
+      - tox-base:
+          requires:
+            - "unit-tests"
+          matrix:
+            parameters:
+              python_version_major: [3]
+              python_version_minor: [8, 9, 10, 11, 12]
```
Comment (kptkin) [inline .circleci/config.yml:1773]: why do you need to test it on all the python versions?

File: tox.ini [hunk 8]
```diff
@@ -184,6 +184,37 @@ commands_pre=
...
+passenv=
+    {[base]passenv}
+    CI_PYTEST_PARALLEL
+    CI
+allowlist_externals=
+    mkdir
+    nox
+    go
+commands_pre=
+    notebooks: ipython kernel install --user --name=wandb_python
```
Comment (kptkin) [inline tox.ini:210]: why do you need this line?

File: .circleci/config.yml (hunk 7 above)
Comment (andrewtruong) [reply .circleci/config.yml:1773]: We can do just 3.8 if it's too expensive

File: tox.ini [hunk 9]
```diff
@@ -184,6 +184,31 @@ commands_pre=
 commands=
     yea {env:CI_PYTEST_SPLIT_ARGS:} --strict --shard {env:SHARD:default} run {posargs:--all}
 
+[testenv:reports-py{38,39,310,311,312}]
+deps=
+    -r{toxinidir}/requirements_test.txt
```
Comment (kptkin) [inline tox.ini:189]: do you need all the deps in requirement_test? i wonder if you can save some time

File: tox.ini [hunk 10]
```diff
@@ -184,6 +184,31 @@ commands_pre=
 commands=
     yea {env:CI_PYTEST_SPLIT_ARGS:} --strict --shard {env:SHARD:default} run {posargs:--all}
 
+[testenv:reports-py{38,39,310,311,312}]
```
Comment (kptkin) [inline tox.ini:187]: i don't think you need a new table, you probably can add it to `[testenv:{,core-}{,notebooks-}py{37,38,39,310,311,312}]` just add the relevant  labels

Comment (kptkin) [reply tox.ini:187]: [testenv:{,core-}{,notebooks-, reports-}py{37,38,39,310,311,312}]

File: tox.ini [hunk 11]
```diff
@@ -184,6 +184,31 @@ commands_pre=
...
+deps=
+    -r{toxinidir}/requirements_test.txt
+    wandb[reports]
+    pydantic
+    polyfactory
```
Comment (kptkin) [inline tox.ini:192]: convert to label `reports: ...`

**Review by dmitryduev — APPROVED** [review]
🥇

**Pull request merged**